### PR TITLE
pre-commit hook + ruff usage

### PR DESCRIPTION
**Pull request by cpelley**
- [ruff](https://docs.astral.sh/ruff/) usage (setup to be equivalent to flake8 + isort + black), as per dagruner, improver_suite etc.
  - Performance improvements from its use.
  - Standardisation across repositories.
- copyright and init checks also turned into pre-commit checks.
  - No longer any need `improver_tests/test_source_code.py` (removed in this PR).
  - `__init__.py` file checks handled via `init_check` script.  Auto-fix supported (creates missing `__init__.py` files).
  - Copyright check will add missing copyright headers.
- pre-commit checking against changed files only in actions.
- Updating `ci.yml` formatting while I was at it to increase readability (line spacing between steps).
- CI workflow running the pre-commit (formally flake8 and black) now only look at files touched by the PR.

Note that you can run pre-commit manually (you needn't wait for a commit):

All files:
```
pre-commit run --all-files
```

Specific files:
```
pre-commit run --files <some-file> <some-other-file>
```

## Issues

- https://github.com/MetOffice/improver_suite/issues/2214

## Why ruff?

This PR switches our use from black, isort, and flake8 to ruff for code formatting and linting. Ruff is gaining popularity due to its performance, efficiency and flexibility. It combines the functionality of black, isort, and flake8 into a single tool (and more if we enable additional rulesets), reducing the complexity of our tooling setup. Ruff offers excellent performance, faster execution times, and the ability to auto-fix issues, ensuring a consistent codebase. Adopting ruff aligns with industry trends and provides a unified approach to code quality with our other repositories.

## Note

Some differences between black and ruff:
https://docs.astral.sh/ruff/formatter/#black-compatibility
(CI now only runs pre-commit on files changed so changes will be made to files as people touch them)

### CODE REVIEW

**Comment by cpelley** [comment]
May be of interest to you @s-boardman - towards consistency between repositories.

**Review by mo-robert-purvis — APPROVED** [review]

**Comment by cpelley** [comment]
@gavinevans - you happy for us to proceed with this?  I can chase-up a second review from the technical team.  I just wasn't sure whether science would be happy with leaving this as a purely technical team concern change.
cheers

**Comment by cpelley** [comment]
I have updated the code style guide to provide more context.

File: pyproject.toml [hunk 1]
```diff
@@ -1,9 +1,9 @@
-[tool.black]
-target-version = ['py36']
+[tool.ruff]
+target-version = "py36"
```
Comment (SamGriffithsMO) [inline pyproject.toml:2]: py36 isn't a valid option and looks like it never will be - https://github.com/astral-sh/ruff/issues/3826#issuecomment-1794880887

Comment (cpelley) [reply pyproject.toml:2]: good spot

**Review by SamGriffithsMO — DISMISSED** [review]
A few more points, I have been testing improver/nbhood/nbhood.py;

1. There appears to be an issue with the copyright_check, it is failing when I don't think it should. Can you add a fix option to copyright_check so users don't have to figure out what the correct header is? (this will also tell me if it is working correctly)
2. bin/improver_tests still contains references to old tools-black/isort/flake8 (is this used/ can it be removed? Or does it need updating)
3. This change appears to update the line-length from 100 down to 88 (not total clear to me why) - note, I am okay with it going to 88. Best guess for cause is the flake8 configuration in setup.cfg (which is still there and presumably could be removed?)
4. There is a formatting change that I think are actually quite bad (for readability), can we do anything about this. Specifically spacing in explicit arrays, e.g. 
```
-                    [[[ 0.75,  0.75,  0.5 ,  0.5 ,  0.5 ,  0.75,  0.75],
-                      [ 0.75,  0.55,  0.55,  0.5 ,  0.55,  0.55,  0.55],
-                      [ 0.55,  0.55,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ],
-                      [ 0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ],
-                      [ 0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.5 ,  0.55,  0.55],
-                      [ 0.55,  0.55,  0.55,  0.5 ,  0.55,  0.55,  0.75],
-                      [ 0.75,  0.75,  0.5 ,  0.5 ,  0.5 ,  0.75,  0.75]]],
+                    (
+                        [
+                            [
+                                [0.75, 0.75, 0.5, 0.5, 0.5, 0.75, 0.75],
+                                [0.75, 0.55, 0.55, 0.5, 0.55, 0.55, 0.55],
+                                [0.55, 0.55, 0.5, 0.5, 0.5, 0.5, 0.5],
+                                [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5],
+                                [0.5, 0.5, 0.5, 0.5, 0.5, 0.55, 0.55],
+                                [0.55, 0.55, 0.55, 0.5, 0.55, 0.55, 0.75],
+                                [0.75, 0.75, 0.5, 0.5, 0.5, 0.75, 0.75],
+                            ]
+                        ],
+                    )
 ```

**Comment by cpelley** [comment]
> There appears to be an issue with the copyright_check, it is failing when I don't think it should.

Can you give an example of this case that failed when you think it shouldn't have?

> Can you add a fix option to copyright_check so users don't have to figure out what the correct header is?

The copyright check will already add a copyright header automatically if it is missing.  When it appears to be present but isn't correct, it's very difficult to automatically fix or present a diff in the general case.  Essentially, checking whether it 'appears' to have a copyright defined is achieved by searching for the term 'copyright' in a comment line.  Beyond that, comments can include arbitrary information so there is no sure way to know where this incorrect copyright finishes.

> bin/improver_tests still contains references to old tools-black/isort/flake8 (is this used/ can it be removed? Or does it need updating)

I'll grep for references to flake8 and black 👍 

> This change appears to update the line-length from 100 down to 88 (not total clear to me why) - note, I am okay with it going to 88. Best guess for cause is the flake8 configuration in setup.cfg (which is still there and presumably could be removed?)

Good spotting, yes I didn't think to look at setup.cfg for the flake8 config.
I'll remove and update the ruff limit to 100 (for now at least).

> There is a formatting change that I think are actually quite bad (for readability), can we do anything about this. Specifically spacing in explicit arrays, e.g.

Yes, I'm not sure.  Will look into it.  I think ideally this would be done with an in-line rule exclusion for such things.

**Comment by cpelley** [comment]
Pulled back into draft to address some points raised by @SamGriffithsMO (thanks for taking a close look at this).

**Comment by SamGriffithsMO** [comment]
> Can you give an example of this case that failed when you think it shouldn't have?

I just ran it on nbhood/nbhood.py, it already has a header that, to me, looks like it should pass. Let me know if you can't reproduce it

```
(improver_production) [sgriffit@vld456:/net/home/h03/sgriffit/repos/improver]$ git status
HEAD detached at origin/PRE_COMMIT
nothing to commit, working tree clean
(improver_production) [sgriffit@vld456:/net/home/h03/sgriffit/repos/improver]$ pre-commit run -v --files improver/nbhood/nbhood.py
ruff.....................................................................Failed
- hook id: ruff
- duration: 0.12s
- exit code: 1

improver/nbhood/nbhood.py:320:89: E501 Line too long (99 > 88)
improver/nbhood/nbhood.py:329:89: E501 Line too long (89 > 88)
improver/nbhood/nbhood.py:331:89: E501 Line too long (93 > 88)
improver/nbhood/nbhood.py:332:89: E501 Line too long (96 > 88)
improver/nbhood/nbhood.py:336:89: E501 Line too long (91 > 88)
improver/nbhood/nbhood.py:338:89: E501 Line too long (89 > 88)
improver/nbhood/nbhood.py:347:89: E501 Line too long (95 > 88)
improver/nbhood/nbhood.py:349:89: E501 Line too long (93 > 88)
improver/nbhood/nbhood.py:352:89: E501 Line too long (92 > 88)
improver/nbhood/nbhood.py:353:89: E501 Line too long (92 > 88)
improver/nbhood/nbhood.py:381:89: E501 Line too long (99 > 88)
improver/nbhood/nbhood.py:382:89: E501 Line too long (96 > 88)
improver/nbhood/nbhood.py:388:89: E501 Line too long (96 > 88)
improver/nbhood/nbhood.py:642:89: E501 Line too long (92 > 88)
Found 14 errors.

ruff-format..............................................................Failed
- hook id: ruff-format
- duration: 0.05s
- files were modified by this hook

1 file reformatted

Check copyright header...................................................Failed
- hook id: copyright_check
- duration: 0.11s
- exit code: 1

Incorrect Copyright header in 'improver/nbhood/nbhood.py'

Check for missing __init__.py files......................................Passed
- hook id: init_check
- duration: 0.02s
```

> Since the `improver_tests/test_source_code.py:test_py_licence` that this copyright check replaces doesn't provide this diff, I'm going to consider it out of scope for this PR.

👍

**Comment by cpelley** [comment]
@nivnac, you may be interested to look at this.

In summary, this is motivated by standardising formatting methods and testing across our repositories.  Also:

1. Smaller and Cleaner Production Environment.**: By removing developer tools like `ruff` from the package's core dependencies, the environment remains lightweight and free from unnecessary developer tools.

2. **Easier Dependency Management**: When developer-only dependencies are managed separately (e.g., using `pre-commit`), it becomes easier to maintain and update them without affecting the conda environment.  This reduces the risk of conflicts between package dependencies and development tools.

3. **Improved Reproducibility**: A clear separation between core and development dependencies ensures that production environments are consistent and reproducible. Development tools can introduce updates or changes that aren’t relevant to production, and by keeping them separate, you reduce potential variability across environments.

4. **Flexibility for Developers**: Pre-commit hooks allow each developer to manage their local development environment while keeping the project’s core environment isolated. For example, developers can update or modify their linting tools without altering the package dependencies.

By keeping developer dependencies separate, we can streamline both development and deployment processes, reducing complexity and ensuring a clear distinction between development tools and production requirements.

Note that IMPROVER uses a number of other developer packages and this is likely just the first step.

File: improver/nbhood/nbhood.py [hunk 2]
```diff
@@ -521,33 +520,43 @@ def pad_and_unpad_cube(self, slice_2d: Cube, kernel: ndarray) -> Cube:
...
-                     [1., 1., 1.]]
+                    [
+                        [
+                            1.0,
+                            1.0,
+                            1.0,
+                        ],
+                        [1.0, 0.0, 1.0],
+                        [1.0, 1.0, 1.0],
+                    ]
```
Comment (SamGriffithsMO) [inline improver/nbhood/nbhood.py:531]: Only nbhood.py seems to have been upset by the switch to ruff. I think reverting these changes and adding `# fmt: off` and `# fmt: on` around these arrays will be sufficient

Comment (SamGriffithsMO) [inline improver/nbhood/nbhood.py:531]: Only nbhood.py seems to have been upset by the switch to ruff. I think reverting these changes and adding `# fmt: off` and `# fmt: on` around these arrays will be sufficient

File: improver/nbhood/nbhood.py [hunk 3]
```diff
@@ -521,33 +520,43 @@ def pad_and_unpad_cube(self, slice_2d: Cube, kernel: ndarray) -> Cube:
...
-                     [ 0.75,  0.75,  1.  ,  0.5 ,  1.  ,  0.75,  0.75]]
+                    [
+                        [0.75, 0.75, 1.0, 0.5, 1.0, 0.75, 0.75],
+                        [0.75, 0.75, 1.0, 0.5, 1.0, 0.75, 0.75],
+                        [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
+                        [0.5, 0.5, 1.0, 0.0, 1.0, 0.5, 0.5],
+                        [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
+                        [0.75, 0.75, 1.0, 0.5, 1.0, 0.75, 0.75],
+                        [0.75, 0.75, 1.0, 0.5, 1.0, 0.75, 0.75],
+                    ]
```
Comment (SamGriffithsMO) [inline improver/nbhood/nbhood.py:559]: another one

File: improver/nbhood/nbhood.py [hunk 4]
```diff
@@ -599,19 +608,23 @@ def pad_and_unpad_cube(self, slice_2d: Cube, kernel: ndarray) -> Cube:
...
+                                [0.75, 0.75, 0.5, 0.5, 0.5, 0.75, 0.75],
+                                [0.75, 0.55, 0.55, 0.5, 0.55, 0.55, 0.55],
+                                [0.55, 0.55, 0.5, 0.5, 0.5, 0.5, 0.5],
+                                [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5],
+                                [0.5, 0.5, 0.5, 0.5, 0.5, 0.55, 0.55],
+                                [0.55, 0.55, 0.55, 0.5, 0.55, 0.55, 0.75],
+                                [0.75, 0.75, 0.5, 0.5, 0.5, 0.75, 0.75],
+                            ]
+                        ],
+                    )
```
Comment (SamGriffithsMO) [inline improver/nbhood/nbhood.py:623]: another

File: improver/nbhood/nbhood.py [hunk 5]
```diff
@@ -599,19 +608,23 @@ def pad_and_unpad_cube(self, slice_2d: Cube, kernel: ndarray) -> Cube:
...
 
             5. The padding is then removed to give::
 
-                   [[[ 0.5,  0.5,  0.5],
-                     [ 0.5,  0.5,  0.5],
-                     [ 0.5,  0.5,  0.5]]]
+                   [[[0.5, 0.5, 0.5], [0.5, 0.5, 0.5], [0.5, 0.5, 0.5]]]
         """
```
Comment (SamGriffithsMO) [inline improver/nbhood/nbhood.py:628]: The final one

**Review by SamGriffithsMO — CHANGES_REQUESTED** [review]
Looking good 👍 , just need to give some special attention to nbhood/nbhood.py and good to go

File: pyproject.toml [hunk 6]
```diff
@@ -1,9 +1,10 @@
...
 
-[tool.isort]
-multi_line_output = 3
-include_trailing_comma = true
-force_grid_wrap = 0
-use_parentheses = true
-line_length = 88
+[tool.ruff.lint]
+extend-select = ["E", "F", "W", "I"]  # add C90 later
+ignore = ["E203", "E731", "E501", "E741"]  # remove "E501", "E741" later
```
Comment (cpelley) [inline pyproject.toml:6]: Ideally we should work towards removing any exclusions.  These exist here now so as not to attempt to cause too many changes (raising standards) whilst first adopting ruff at the same time.

**Review by SamGriffithsMO — APPROVED** [review]
👍

**Review by mo-robert-purvis — APPROVED** [review]
That was a lot of files!

File: README.md [hunk 7]
```diff
@@ -42,3 +42,10 @@ activate your new improver environment
 ```
 conda activate improver
 ```
+
+## Pre-commit Hook
+OPTIONAL: A pre-commit hook can be added to facilitate the development of this code base.
+Ensure that you have python available on the path, then install the pre-commit hook by running `pre-commit install` from within your working copy.
+pre-commit checks will run against modified files when you commit from then on.
+
+These pre-commit hooks will run as part of continuous integration to maintain standard in the project.
```
Comment (bayliffe) [inline README.md:51]:
```suggestion
These pre-commit hooks will run as part of continuous integration to maintain standards in the project.
```

**Review by bayliffe — COMMENTED** [review]
Will you remake the improver_production environment to exclude the now-redundant packages (black etc), or does the environment need to be left unchanged until post-PS46 to leave things genuinely unchanged?

One tiny typo highlighted. I've not looked at every file, but I get the gist of it (thanks to Sam for being more rigorous). Happy with this change, thanks.

As an aside, I note that the copyright check is by far the slowest of the tests to run, so we might optimise that in future (not required here) to save ourselves whole seconds.

**Comment by cpelley** [comment]
> Will you remake the improver_production environment to exclude the now-redundant packages (black etc), or does the environment need to be left unchanged until post-PS46 to leave things genuinely unchanged?

Keep our existing environments unchanged 👍 
Whilst there are now some redundant packages within, I don't think worth the time/effort/testing changing our pre-existing environments.

> One tiny typo highlighted. I've not looked at every file, but I get the gist of it (thanks to Sam for being more rigorous). Happy with this change, thanks.

Done.  Thanks.

> As an aside, I note that the copyright check is by far the slowest of the tests to run, so we might optimise that in future (not required here) to save ourselves whole seconds.

Whilst true that the copyright header check will be order of magnitudes slower than ruff, it will likely be faster than the pre-existing python copyright header checker that it was replacing.  Also, whilst checking the entire repository might take seconds, the pre-commit only runs against files that have change compared to master.  For this branch, I forcibly ran the pre-commit against on all files since we are changing the the formatting testing approach.  Under normal circumstances, a development branch will be changing tens of files at most (normally), so I think we are OK??

**Comment by cpelley** [comment]
> @nivnac, you may be interested to look at this.
> 
> ...

@nivnac, you happy for us to merge this?

**Review by bayliffe — COMMENTED** [review]
Thanks @cpelley the copyright check is noticeably much faster.

**Review by bayliffe — DISMISSED** [review]

File: README.md [hunk 8]
```diff
@@ -42,3 +42,10 @@ activate your new improver environment
 ```
 conda activate improver
 ```
+
+## Pre-commit Hook
+OPTIONAL: A pre-commit hook can be added to facilitate the development of this code base.
+Ensure that you have python available on the path, then install the pre-commit hook by running `pre-commit install` from within your working copy.
+pre-commit checks will run against modified files when you commit from then on.
+
+These pre-commit hooks will run as part of continuous integration to maintain standards in the project.
```
Comment (mo-robert-purvis) [inline README.md:51]: Which standards? Should this say "code quality standards"?

File: copyright_check [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                echo -e ${copyright_header} >> ${tmp_file}
+                tail -n +2 ${filepath} >> ${tmp_file}
+            else
+                echo -e ${copyright_header} > ${tmp_file}
+                cat ${filepath} >> ${tmp_file}
+            fi
+                mv ${tmp_file} ${filepath}
+                found_error=1
+            fi
+        fi
```
Comment (mo-robert-purvis) [inline copyright_check:118]: Reduce indentation by one level.
```suggestion
            fi
            mv ${tmp_file} ${filepath}
            found_error=1
        fi
    fi
```

**Review by mo-robert-purvis — DISMISSED** [review]
Some small cosmetic changes suggested

File: README.md (hunk 8 above)
Comment (cpelley) [reply README.md:51]: 🤷‍♂️ I don't see what might be ambiguous myself.
Making the below suggested change to I'll include this one too anyway.

**Review by mo-robert-purvis — APPROVED** [review]

**Review by SamGriffithsMO — APPROVED** [review]

**Review by nivnac — APPROVED** [review]

**Pull request merged**